src/test/java/com/epam/idea/rest/controller/TagControllerTest.java

```diff
@@ -0,0 +1,109 @@
+import com.epam.idea.rest.config.TestConfig;
+import com.epam.idea.rest.config.WebAppConfig;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.hateoas.Link;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+import org.springframework.test.context.web.WebAppConfiguration;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+import static com.epam.idea.rest.TestUtils.APPLICATION_JSON_UTF8;
+import static com.epam.idea.rest.resource.support.JsonPropertyName.ID;
+import static java.util.Arrays.asList;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+/**
+ * Created by Ihar_Niakhlebau on 18-Mar-15.
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes = {TestConfig.class, WebAppConfig.class})
+@WebAppConfiguration
+public class TagControllerTest {
+
+	@Autowired
+	private TagService tagServiceMock;
+
+	@Autowired
+	private WebApplicationContext webApplicationContext;
+
+	private MockMvc mockMvc;
+
+	@Before
+	public void setUp() throws Exception {
+		Mockito.reset(tagServiceMock);
+		mockMvc = MockMvcBuilders.webAppContextSetup(this.webApplicationContext).build();
+	}
+
+	@Test
+	public void shouldReturnAllFoundTags() throws Exception {
+		Tag foundTag = TestTagBuilder.aTag().build();
+		when(tagServiceMock.findAll()).thenReturn(asList(foundTag));
+		mockMvc.perform(get("/api/v1/tags")
+				.accept(APPLICATION_JSON_UTF8))
+				.andExpect(status().isOk())
+				.andExpect(content().contentType(APPLICATION_JSON_UTF8))
+				.andExpect(jsonPath("$[0]." + ID).value(is(((int) foundTag.getId()))))
+				.andExpect(jsonPath("$[0].name").value(is(foundTag.getName())))
+				.andExpect(jsonPath("$[0].links[0].rel").value(is(Link.REL_SELF)))
+				.andExpect(jsonPath("$[0].links[0].href").value(containsString("/api/v1/tags/" + foundTag.getId())))
+				.andExpect(jsonPath("$[0].links", hasSize(1)));

```

Comment:
assert first that `links` array has size equals 1 and only after that u can retrieve 1-st element otherwise if `links` is empty u will get irrelevant to the test error message